R/ms-sql.R

```diff
@@ -0,0 +1,227 @@
+        )
+      )
+      if (nrow(res) > 1) {
+        stop("Corrupted cache: more than one entry for ", key)
+      }
+      nrow(res) == 1
+    },
+    #' @description
+    #' Get a key from the cache
+    #' @param key Name of the key.
+    #' @return The value stored using the `key`
+    # Inspied by @jrosell https://stackoverflow.com/a/70288183/8236642
+    get = function(key) {
+      if (self$has_key(key)) {
+        tmp <- tempfile(fileext = ".RDS")
+        on.exit({
+          unlink(tmp, TRUE, TRUE)
+        })
+        out <- DBI::dbGetQuery(
+          private$interface,
+          sprintf(
+            "SELECT * FROM %s WHERE id = '%s';",
+            private$cache_table,
+            key
+          )
+        )
+
+        # Handling the case where the value has been deleted in-between
+        # (should be very, very corner case)
+        res <- tryCatch(
+          {
+            unserialized_out <- unserialize(out$cache[[1]])
+            writeBin(object = unserialized_out, con = tmp)
+            readRDS(tmp)
+          },
+          error = function(e) {
+            return(NULL)
+          }
+        )
+        if (is.null(res)) {
+          return(
+            structure(list(), class = "key_missing")
+          )
+        }
+        return(
+          res
+        )
+      } else {
+        return(
+          structure(list(), class = "key_missing")
+        )
+      }
+    },
+    #' @description
+    #' Set a key in the cache
+    #' @param key Name of the key.
+    #' @param value Value to store
+    #' @return Used for side effect
+    set = function(key, value) {
+      if (!self$has_key(key)) {
+        temp_file <- tempfile(fileext = ".RDS")
+        saveRDS(value, file = temp_file)
+        pdf <- readBin(con = temp_file, what = raw(), n = file.info(temp_file)$size)
+        pdf_serialized <- serialize(pdf, NULL)
+
+        DBI::dbWriteTable(
+          private$interface,
+          private$cache_table,
+          data.frame(
+            id = key,
+            cache = I(list(pdf_serialized))
+          ),
+          append = TRUE
+        )
+      }
+    },
+    #' @description
+    #' Clear all the cache
+    #' @return Used for side-effect
+    reset = function() {
+      DBI::dbRemoveTable(
+        private$interface,
+        private$cache_table
+      )
+      DBI::dbCreateTable(
+        private$interface,
+        private$cache_table,
+        fields = c(
+          id = "varchar(max)",
+          cache = "varbinary(max)"
+        )
+      )
+    },
+    #' @description
+    #' Remove a key/value pair
+    #' @param key Name of the key.
+    #' @return Used for side-effect
+    remove = function(key) {
+      DBI::dbExecute(
+        private$interface,
+        sprintf(
+          "DELETE FROM %s WHERE id = '%s';",
+          private$cache_table,
+          key
+        )
+      )
+    },
+    #' @description
+    #' List all the keys in the cache
+    #' @return A list of keys
+    keys = function() {
```

Comment:
`ms-sql$keys` is the same as `postgres$keys` so there is no need to redefine here (will be passed through inheritance).